File: 01-clases-objetos/python/perro.py
"""
perro.py - LA CLASE (el molde)

DIFERENCIA CON JAVA:
  Java OBLIGA a que el archivo se llame igual que la clase publica.
  Python NO obliga a nada: el archivo puede llamarse como quieras y
  puede contener varias clases.

  Pero la costumbre (PEP 8) es:
      archivo  ->  minusculas_con_guion_bajo   perro.py, cuenta_bancaria.py
      clase    ->  CamelCase                   Perro,   CuentaBancaria

Este archivo no se ejecuta solo: solo define el molde.
"""


class Perro:

    # CONSTRUCTOR: se ejecuta al crear cada objeto
    # En Python los ATRIBUTOS no se declaran arriba: nacen aqui, con self.
    def __init__(self, nombre, raza, edad):
        self.nombre = nombre
        self.raza = raza
        self.edad = edad

    # METODOS: siempre reciben "self" como primer parametro
    def ladrar(self):
        print(f"{self.nombre} dice: Guau!")

    def cumplir_anios(self):
        self.edad = self.edad + 1
        print(f"{self.nombre} ahora tiene {self.edad} anios.")

    def presentarse(self):
        print(f"Soy {self.nombre}, un {self.raza} de {self.edad} anios.")

File: 01-clases-objetos/java/Perro.java
/*
 * Perro.java - LA CLASE (el molde)
 *
 * REGLA DE ORO DE JAVA:
 *   un archivo .java = una clase publica, y el nombre del archivo
 *   tiene que ser EXACTAMENTE igual al de la clase.
 *   Perro.java  ->  public class Perro
 *
 * Esta clase no se ejecuta sola: no tiene main(). Es solo el molde.
 */
public class Perro {

    // ATRIBUTOS: los datos que cada perro tiene por separado
    public String nombre;
    public String raza;
    public int edad;

    // CONSTRUCTOR: se ejecuta al crear cada objeto con "new"
    // "this.nombre" = el atributo del objeto.  "nombre" = el parametro.
    public Perro(String nombre, String raza, int edad) {
        this.nombre = nombre;
        this.raza = raza;
        this.edad = edad;
    }

    // METODOS: las cosas que el perro sabe hacer
    public void ladrar() {
        System.out.println(this.nombre + " dice: Guau!");
    }

    public void cumplirAnios() {
        this.edad = this.edad + 1;
        System.out.println(this.nombre + " ahora tiene " + this.edad + " anios.");
    }

    public void presentarse() {
        System.out.println("Soy " + nombre + ", un " + raza + " de " + edad + " anios.");
    }
}
